docs/framework-design.md

```diff
@@ -0,0 +1,145 @@
+* Library source files must not rely on headers from the framework.
+
+#### Requirements for users of consuming repositories
+
+Corresponding to the requirements on the repository above:
+
+* Contributors need the framework submodule.
+* Users who wish to run full CI tests need the framework submodule.
+* Users who want to build or run tests need the `framework` directory content.
+* Users who merely want to build the library, and who have the configuration-independent files already generated, do not need the `framework` directory content.
+
+## Contents of the framework repository
+
+### Criteria for inclusion
+
+In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and branch-specific content; such files should be split into a shared part and a non-shared part.
+
+For example:
+
+* `test_suite_*.function` contains a lot of code that is specific to each branch of each consuming repository. Even when the same test function exists in all maintained branches, there are often minor differences such as a deprecated alternative that only exists in older branches, differences in compile-time dependencies, etc. Thus we do not expect to share these files. Common code can go into separate `.c` files, historically under `tests/src`, that are in the framework repository.
+* `test_suite_*.data` contains many test cases that exist in all maintained branches. However the exact expression of these test cases are often different, for example due to compile-time dependencies. Furthermore the set of test cases is often different, for example due to cryptographic mechanisms that are added or removed. Thus we do not expect to share these files. Where there is a lot of commonality, the test cases can be generated from a script located in the framework repository, with code in the generation script to handle the differing parts.
+
+## CI architecture
+
+Tasks:
+
+* CI in consuming repositories must support Git submodules. Other than that, keep the CI as it is now. In particular, the CI in consuming repositories does not need to consider anything but the commit that the framework submodule points to.
+* CI in the framework repository should run a subset of the CI of all consuming repositories, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming repository to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.
+* When merging a pull request to in an official branch in a consuming repository (`development`, LTS branches), check that the commit.
+
+TODO: once this is set up, detail the processes here.
```

Comment:
```suggestion
```
Just remove this?

Reply:
Er, why? Both sentences are there for a reason. I will finish the unfinished sentence.